### PR TITLE
Add Battle of Yavin and associated SL schema updates

### PR DESCRIPTION
**Pull request by chrisalleng**

### CODE REVIEW

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 1]
```diff
@@ -307,6 +307,262 @@
...
+      "limited": 1,
+      "cost": 5,
+      "standardLoadout": true,
+      "xws": "biggsdarklighter-battleofyavin",
+      "ability": "During the System Phase, you may choose 1 friendly ship at range 1. If you do, treat your initiative as equal to the chosen ship's initiative until the end of the Activation Phase.",
+      "shipAbility": {
+        "name": "Hope",
+        "text": "After another friendly ship at range 0-3 is destroyed, you may perform a [Focus] or [Boost] action."
+      },
+      "upgrades": [
```
Comment (chrisalleng) [inline data/pilots/rebel-alliance/t-65-x-wing.json:324]: The "upgrades" name is totally new. Any issue with upgrades? I couldn't think of anything better.

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 2]
```diff
@@ -307,6 +307,262 @@
...
+        {
+          "title": "Attack Speed",
+          "slots": ["Talent"],
+          "ability": "After you fully execute a [3 [Straight]] or [4 [Straight]] maneuver, you may perform a boost using the [1 [Straight]] template. (This is not an action)."
+        },
+        {
+          "title": "Selfless",
```
Comment (chrisalleng) [inline data/pilots/rebel-alliance/t-65-x-wing.json:331]: Some of these upgrades already exist, I figured it made more sense to just copy the text here since there's no guarantee in the long term that the text on these cards will match any upgrade. In some cases we know it doesn't, but I'm not sure if there's any non-limited upgrades where that is true.

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 3]
```diff
@@ -307,6 +307,262 @@
...
+          "slots": ["Talent"],
+          "ability": "While another friendly ship at range 0-1 defends, before the Neutralize Results step, if you are in the attack arc, you may suffer 1 [Critical Hit] damage to cancel 1 [Critical Hit] result.",
+          "limited": 1
```
Comment (chrisalleng) [inline data/pilots/rebel-alliance/t-65-x-wing.json:334]: Selfless has a limited dot on Bigg's card. That seems like an accident? Haven't seen any discussion on it anywhere, but also have had my head under a rock. Anyone know if this is supposed to be limited?

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 4]
```diff
@@ -307,6 +307,262 @@
       "extended": true,
       "keywords": ["X-wing"],
       "epic": true
+    },
```
Comment (chrisalleng) [inline data/pilots/rebel-alliance/t-65-x-wing.json:310]: Pilots seems to be alphabetically sorted in their respective chassis. I was considering making another section entirely of standardpilots or something, but I'm worried about compatibility with whatever xws parsers people may have already written.

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 5]
```diff
@@ -307,6 +307,262 @@
...
+    {
+      "name": "Biggs Darklighter",
+      "caption": "Battle of Yavin",
+      "initiative": 3,
+      "limited": 1,
+      "cost": 5,
+      "standardLoadout": true,
```
Comment (chrisalleng) [inline data/pilots/rebel-alliance/t-65-x-wing.json:317]: standardLoadout seemed like a reasonable choice name for this label on these ships to me, but open to better suggestions as well. Presumably we need one?

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 6]
```diff
@@ -307,6 +307,262 @@
...
+        },
+        {
+          "title": "R2-F2",
+          "slots": ["Astromech"],
+          "ability": "After you reveal your dial, you may spend 1 [Charge] and gain 1 disarm token to recover 1 shield.",
+          "charges": { "value": 2, "recovers": 0 },
+          "limited": 1
+        }
+      ],
+      "slots": [
```
Comment (chrisalleng) [inline data/pilots/rebel-alliance/t-65-x-wing.json:357]: Having an empty set of slots feels weird, but again thought it was probably better to include for compatibility.

File: data/pilots/rebel-alliance/t-65-x-wing.json [hunk 7]
```diff
@@ -307,6 +307,262 @@
...
+      ],
+      "image": "https://infinitearenas.com/xw2/images/quickbuilds/biggsdarklighter-battleofyavin.png",
+      "artwork": "https://infinitearenas.com/xw2/images/artwork/pilots/biggsdarklighter.png",
```
[the remark this answers is not in the document]
Comment (danrs) [reply data/pilots/rebel-alliance/t-65-x-wing.json:360]: Feel free to use whatever art is convenient/good, but I don't think it's worth spending too much time on. I think we still have some missing art on pilots that are pointing at ffg, etc, so updating our art probably deserves its own ticket. The correct solution might be to have infinitearenas host the nicer arts and then just run a script to autogenerate the art field in this repo.

File: data/pilots/rebel-alliance/t-65-x-wing.json (hunk 2 above)
Comment (danrs) [reply data/pilots/rebel-alliance/t-65-x-wing.json:331]: Putting the text of every upgrade on every pilot is too much repetition. Instead, I'd do [what yasb does](https://github.com/raithos/xwing/blob/master/coffeescripts/content/cards-common.coffee).

1: Add new upgrade entries for upgrades which are unique to standardized pilots. This means things like attack run or yavin chewie crew. Upgrades which appear unchanged on standardized loadout pilots should not be duplicated.

Upgrades from standardized loadout cards should:
- have the name of the expansion in the card id (`foo-battleofyavin`)
- have no cost field (although leaving cost as 0 is fine)
- have a new field `standardLoadoutOnly: true`. This should be its own field, it's logically distinct from `restrictions`

This will require updates to the upgrade schema.

2: list upgrade IDs in a pilot's `upgrades` field

**Review by danrs — CHANGES_REQUESTED** [review]

File: data/pilots/rebel-alliance/t-65-x-wing.json (hunk 6 above)
Comment (chrisalleng) [reply data/pilots/rebel-alliance/t-65-x-wing.json:357]: This makes sense. Ironically, we'll still have issues with ships with empty slots (generic empire TIE Fighters all have 0 loadout and no upgrade bar), but I think it's clear what's going on at that point then.

File: data/pilots/rebel-alliance/t-65-x-wing.json (hunk 2 above)
Comment (chrisalleng) [reply data/pilots/rebel-alliance/t-65-x-wing.json:331]: Came back to make effectively the same comment, that yes this duplication is just too much, even if it means some oddities, like for example if the Selfless on Biggs doesn't get an errata (though I'm not even sure what that would mean).

File: data/pilots/rebel-alliance/t-65-x-wing.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisalleng) [reply data/pilots/rebel-alliance/t-65-x-wing.json:360]: Somewhere on my life TODO list is write a script to point all the art/images over to infinite arenas. #521 already refers to it. I'll get these using the infinite arenas work, and not stress about the rest for now.

File: data/pilots/rebel-alliance/t-65-x-wing.json (hunk 6 above)
Comment (danrs) [reply data/pilots/rebel-alliance/t-65-x-wing.json:357]: Yep, empty slots is valid and means "this pilot has no upgrade slots". That's different to `standardLoadout` pilots which ideally should have no slots field, because the concept of slots isn't valid on that kind of pilot.

If the conditional stuff in the schema is too difficult then we can just accept an empty slots entry for standardLoadout pilots, but the conditional stuff would be really helpful to future contributors.

File: data/upgrades/astromech.json [hunk 8]
```diff
@@ -651,5 +651,129 @@
     "standard": true,
     "extended": true,
     "epic": true
+  },
+  {
+    "name": "R2-A3",
+    "limited": 1,
+    "xws": "r2a3-battleofyavin",
```
Comment (chrisalleng) [inline data/upgrades/astromech.json:658]: I generally put the -battleofyavin suffix on these upgrades and pilots, even when no "regular" version exists already. I think this breaks xws norms, but YASB is already doing this, and I think it makes sense for this style of card to have the suffix on all cards for clarity. Any strong opinions?

File: data/upgrades/modification.json [hunk 9]
```diff
@@ -535,5 +535,24 @@
     "standard": true,
     "extended": true,
     "epic": true
+  },
+  {
+    "name": "Unstable Sublight Engines",
+    "limited": 0,
+    "xws": "unstablesublightengines",
```
Comment (chrisalleng) [inline data/upgrades/modification.json:542]: TODO here - forgot -battleofyavin suffix, clean up when either we decide all of these cards/abilities should have this tag, or the new names shouldn't have it.

File: data/upgrades/modification.json [hunk 10]
```diff
@@ -535,5 +535,24 @@
...
+    "sides": [
+      {
+        "ability": "After you are destroyed, you must execute a [1 [Straight]] maneuver, then each other ship at range 0-1 suffers 1 [Hit] damage.",
+        "title": "Unstable Sublight Engines",
+        "type": "Modification",
+        "slots": ["Modification"],
+        "charges": { "value": 1, "recovers": 0 }
```
Comment (chrisalleng) [inline data/upgrades/modification.json:549]: shouldn't have charges

File: data/upgrades/modification.json [hunk 11]
```diff
@@ -535,5 +535,24 @@
...
+      }
+    ],
+    "cost": { "value": 6 },
```
Comment (chrisalleng) [inline data/upgrades/modification.json:552]: shouldn't have cost

File: data/upgrades/astromech.json [hunk 12]
```diff
@@ -651,5 +651,129 @@
...
+    "sides": [
```
Comment (chrisalleng) [inline data/upgrades/astromech.json:659]: Having sides on these abilities feels weird. Compatibility with other cards probably makes more sense than having the schema support two types of cards/abilities?

File: data/upgrades/astromech.json [hunk 13]
```diff
@@ -651,5 +651,129 @@
...
+    "epic": true,
+    "standardLoadoutOnly": true
+  },
+  {
+    "name": "Precise Astromech",
+    "limited": 0,
+    "xws": "preciseastromech-battleofyavin",
+    "sides": [
+      {
+        "ability": "After you perform an action, you may spend 1 [Charge] to perform a red [Lock] action.",
```
Comment (chrisalleng) [inline data/upgrades/astromech.json:751]: needs charges

File: data/upgrades/astromech.json (hunk 12 above)
Comment (danrs) [reply data/upgrades/astromech.json:659]: Yes, keep sides to preserve compatibility.

File: data/upgrades/astromech.json (hunk 8 above)
Comment (danrs) [reply data/upgrades/astromech.json:658]: Put it on. Aligning with yasb streamlines things, and I was leaning that way already.

File: tests/schemas/pilot.schema.json [hunk 14]
```diff
@@ -10,6 +10,13 @@
     "xws": { "type": "string", "pattern": "^[a-z0-9-]+$" },
     "ability": { "type": "string" },
     "text": { "type": "string" },
+    "standardLoadout": { "type": "boolean"},
+    "upgrades": {
+      "type": "array",
+      "items": {
+        "type": "string"
```
Comment (chrisalleng) [inline tests/schemas/pilot.schema.json:17]: Is there a way we can point this towards the list of upgrades, to validate that all upgrades exist in upgrades/*.json?

File: data/pilots/rebel-alliance/t-65-x-wing.json (hunk 6 above)
Comment (chrisalleng) [reply data/pilots/rebel-alliance/t-65-x-wing.json]: I updated the schema to accept upgrades, but haven't done the oneOf for upgrades/slots. I'll get it in the first round of typo fixes.

File: tests/schemas/pilot.schema.json [hunk 15]
```diff
@@ -125,6 +132,75 @@
       "required": ["name", "text"],
       "additionalProperties": false
     },
+    "shipStats": {
```
Comment (chrisalleng) [inline tests/schemas/pilot.schema.json:135]: this is literally a direct copy/paste of the "stats" schema for ships. Makes me think there might be a better way to do this, but presumably any stat could be updated in the same way.

Comment (danrs) [reply tests/schemas/pilot.schema.json:135]: I think it's an appropriate solution for now - if we change this in future we should change `shipActions` as well

File: tests/schemas/pilot.schema.json (hunk 14 above)
Comment (danrs) [reply tests/schemas/pilot.schema.json:17]: I don't think this can be done in the schema, because you'd first have to load the list of upgrades into memory and then compare these entries against them.

Another example of a test too complex to live purely in schema is detection of duplicate XWS ids: https://github.com/guidokessels/xwing-data2/blob/master/tests/xws.test.js

Possible solutions here:
1. no check (not ideal, but error impact is fairly low)
2. check that entries match the rough XWS regex (see line 10, `"xws": { "type": "string", "pattern": "^[a-z0-9-]+$" },`)
3. Write another test like `xws.text.js` which loads all the valid upgrade xws IDs into an array and then matches each of the standardLoadout upgrade Ids against that array

(3) is clearly ideal but involves writing js, and (2) is probably sufficient if you are pressed for time. Up to you which of (2) and (3) you have capacity to implement.

File: package-lock.json [hunk 16]
```diff
@@ -0,0 +1,4948 @@
+{
+  "name": "xwing-data2",
+  "version": "3.3.4",
```
Comment (danrs) [inline package-lock.json:3]: Please update the version to 3.5.0

Also update in data/manifest.json

File: tests/schemas/pilot.schema.json (hunk 14 above)
Comment (danrs) [reply tests/schemas/pilot.schema.json:17]: Actually, after thinking further, just do (2). All the errors I saw in this PR were typing the wrong upgrade, not a typo in the upgrade name. So the regex should be plenty

File: tests/schemas/pilot.schema.json [hunk 17]
```diff
@@ -10,6 +10,13 @@
     "xws": { "type": "string", "pattern": "^[a-z0-9-]+$" },
     "ability": { "type": "string" },
     "text": { "type": "string" },
+    "standardLoadout": { "type": "boolean"},
+    "upgrades": {
```
Comment (danrs) [inline tests/schemas/pilot.schema.json:14]: Please put `upgrades` and `slots` inside a `OneOf` so that only one of them can be present. There's an example of `OneOf` in `shipstats` I think

Comment (danrs) [reply tests/schemas/pilot.schema.json:14]: We also discussed adding some conditional logic around `upgrades` and `standardLoadout`, but I think there might be a cleaner solution. What if we remove the `standardLoadout` boolean and rename the `upgrades` array to `standardLoadout`? That's effectively what we were going for: only ships that have `standardLoadout` have `upgrades`.

So the resulting schema would look something like:
```
...
OneOf{
  "slots": blahblahblah list of slots,
  "standardLoadout": blahblahblah list of upgrades
}
...
```

**Review by danrs — CHANGES_REQUESTED** [review]

File: data/pilots/galactic-empire/tie-advanced-x1.json [hunk 18]
```diff
@@ -204,6 +204,31 @@
...
+      "xws": "darthvader-battleofyavin",
+      "standardLoadout": true,
+      "ability": "While you perform an attack, you may spend 1 [Force] to change 1 blank result to a [Hit] result.",
+      "image": "https://infinitearenas.com/xw2/images/quickbuilds/darthvader-battleofyavin.png",
+      "force": { "value": 3, "recovers": 1, "side": ["dark"] },
+      "shipAbility": {
+        "name": "Advanced Targeting Computer",
+        "text": "While you perform a primary attack against a defender you have locked, roll 1 additional attack die and change 1 [Hit] result to a [Critical Hit] result."
+      },
+      "shipStats": [
```
[the remark this answers is not in the document]
Comment (chrisalleng) [reply data/pilots/galactic-empire/tie-advanced-x1.json:223]: didn't even think of this, good point.

File: tests/schemas/pilot.schema.json (hunk 17 above)
Comment (chrisalleng) [reply tests/schemas/pilot.schema.json:14]: I like combining them, it really is duplicate information. Good idea!

Comment (chrisalleng) [reply tests/schemas/pilot.schema.json]: Because these are both top level properties and not nested inside another, the oneof has to be in the same place as required to my knowledge. Will tag you in the update to make sure it looks correct.

File: tests/schemas/pilot.schema.json [hunk 19]
```diff
@@ -262,5 +262,9 @@
     }
   },
   "required": ["name", "standard", "extended", "epic"],
+  "oneOf": [
```
Comment (chrisalleng) [inline tests/schemas/pilot.schema.json:265]: @danrs here's where the oneof is - makes logical sense next to the required slots, but double checking there isn't a better way to do it.

File: package.json [hunk 20]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "xwing-data2",
-  "version": "3.3.4",
+  "version": "3.5.0",
```
Comment (chrisalleng) [inline package.json:3]: should this also be incremented? I assumed, but not sure how if there's some auto generation or something I have to worry about because JS is the enemy.

File: data/pilots/galactic-empire/tie-in-interceptor.json [hunk 21]
```diff
@@ -264,6 +264,126 @@
...
+      "extended": true,
+      "cost": 6,
+      "charges": { "value": 2, "recovers": 1 },
+      "shipStats": [
+        { "type": "hull", "value": 4 }
+      ],
+      "upgrades": ["predator", "fanatic-battleofyavin"],
+      "ability": "Before a friendly TIE at range 0-1 would suffer damage, you may spend 2 [Charge]. If you do, prevent 1 damage.",
+      "keywords": ["TIE"],
+      "image": "https://infinitearenas.com/xw2/images/quickbuilds/darthvader-battleofyavin.png",
```
[the remark this answers is not in the document]
Comment (danrs) [reply data/pilots/galactic-empire/tie-in-interceptor.json:291]: This still needs correcting: s/darthvader/idenversio/

File: tests/schemas/pilot.schema.json (hunk 19 above)
Comment (danrs) [reply tests/schemas/pilot.schema.json:265]: I think you either do it here or you do it around the fields themselves. This is probably neater as it means less indentation, and if it works I'm fine with it

**Review by danrs — CHANGES_REQUESTED** [review]

**Review by danrs — APPROVED** [review]
